File: src/agent/anthropic.rs
/// Anthropic provider implementation.
///
/// Serializes `Message` IR → Anthropic messages API format, calls via reqwest,
/// and parses the response back to `ToolCallMsg` IR.
use std::time::Duration;

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};

use super::ir::{ContentPart, Message, ToolCallMsg};
use super::provider::{CallResponse, Provider};
use super::tool::ToolDef;

const ANTHROPIC_VERSION: &str = "2023-06-01";
const DEFAULT_TIMEOUT_SECS: u64 = 60;
const DEFAULT_MAX_TOKENS: u32 = 64000;

pub struct AnthropicProvider {
    client: reqwest::Client,
    endpoint: String,
    api_key: String,
    model: String,
    max_tokens: u32,
}

impl AnthropicProvider {
    pub fn new(endpoint: &str, api_key: &str, model: &str) -> Result<Self> {
        let client = reqwest::Client::builder()
            .timeout(Duration::from_secs(DEFAULT_TIMEOUT_SECS))
            .build()
            .context("Failed to create HTTP client")?;

        Ok(Self {
            client,
            endpoint: endpoint.trim_end_matches('/').to_string(),
            api_key: api_key.to_string(),
            model: model.to_string(),
            max_tokens: DEFAULT_MAX_TOKENS,
        })
    }

    pub fn with_max_tokens(mut self, max_tokens: u32) -> Self {
        self.max_tokens = max_tokens;
        self
    }
}

impl Provider for AnthropicProvider {
    fn call<'a>(
        &'a self,
        messages: &'a [Message],
        tools: &'a [ToolDef],
    ) -> std::pin::Pin<Box<dyn std::future::Future<Output = Result<CallResponse>> + Send + 'a>> {
        Box::pin(async move {
            let system = messages.iter().find_map(|m| match m {
                Message::System(text) => Some(serde_json::json!([{
                    "type": "text",
                    "text": text,
                    "cache_control": {"type": "ephemeral"}
                }])),
                _ => None,
            });

            let api_messages = serialize_messages(messages);
            let api_tools = serialize_tools(tools);

            let body = ApiRequest {
                model: self.model.clone(),
                max_tokens: self.max_tokens,
                system,
                messages: api_messages,
                tools: api_tools,
            };

            let url = format!("{}/messages", self.endpoint);
            let resp = self
                .client
                .post(&url)
                .header("x-api-key", &self.api_key)
                .header("anthropic-version", ANTHROPIC_VERSION)
                .header("content-type", "application/json")
                .json(&body)
                .send()
                .await
                .context("Anthropic API request failed")?;

            let status = resp.status();
            let body_text = resp
                .text()
                .await
                .context("Failed to read Anthropic response")?;

            if !status.is_success() {
                if let Ok(err) = serde_json::from_str::<ErrorResponse>(&body_text) {
                    anyhow::bail!("Anthropic API error ({}): {}", status, err.error.message);
                }
                anyhow::bail!("Anthropic API error ({}): {}", status, body_text);
            }

            let response: ApiResponse =
                serde_json::from_str(&body_text).with_context(|| {
                    format!(
                        "Failed to parse Anthropic response: {}",
                        &body_text[..200.min(body_text.len())]
                    )
                })?;

            let tool_calls = response
                .content
                .iter()
                .filter_map(|block| match block {
                    ResponseBlock::ToolUse { id, name, input } => Some(ToolCallMsg {
                        id: id.clone(),
                        name: name.clone(),
                        arguments: input.to_string(),
                    }),
                    _ => None,
                })
                .collect();

            let text: String = response
                .content
                .iter()
                .filter_map(|block| match block {
                    ResponseBlock::Text { text } => Some(text.as_str()),
                    _ => None,
                })
                .collect::<Vec<_>>()
                .join("");
            let text = if text.is_empty() { None } else { Some(text) };

            Ok(CallResponse { tool_calls, text })
        })
    }
}

// ── Anthropic API wire types (private) ──

#[derive(Serialize)]
struct ApiRequest {
    model: String,
    max_tokens: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    system: Option<serde_json::Value>,
    messages: Vec<serde_json::Value>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    tools: Vec<serde_json::Value>,
}

#[derive(Deserialize, Debug)]
struct ApiResponse {
    content: Vec<ResponseBlock>,
    #[allow(dead_code)]
    stop_reason: Option<String>,
}

#[derive(Deserialize, Debug)]
#[serde(tag = "type")]
enum ResponseBlock {
    #[serde(rename = "text")]
    Text {
        #[allow(dead_code)]
        text: String,
    },
    #[serde(rename = "tool_use")]
    ToolUse {
        id: String,
        name: String,
        input: serde_json::Value,
    },
}

#[derive(Deserialize, Debug)]
struct ErrorResponse {
    error: ErrorDetail,
}

#[derive(Deserialize, Debug)]
struct ErrorDetail {
    message: String,
}

// ── IR → Anthropic JSON serialization ──

fn serialize_messages(messages: &[Message]) -> Vec<serde_json::Value> {
    messages
        .iter()
        .filter_map(|msg| match msg {
            Message::System(_) => None, // handled separately
            Message::User(parts) => {
                Some(serde_json::json!({
                    "role": "user",
                    "content": serialize_content_parts(parts)
                }))
            }
            Message::Assistant { text, tool_calls } => {
                let mut blocks: Vec<serde_json::Value> = Vec::new();
                if let Some(text) = text {
                    if !text.is_empty() {
                        blocks.push(serde_json::json!({
                            "type": "text",
                            "text": text
                        }));
                    }
                }
                for tc in tool_calls {
                    let input: serde_json::Value =
                        serde_json::from_str(&tc.arguments).unwrap_or_default();
                    blocks.push(serde_json::json!({
                        "type": "tool_use",
                        "id": tc.id,
                        "name": tc.name,
                        "input": input
                    }));
                }
                Some(serde_json::json!({
                    "role": "assistant",
                    "content": blocks
                }))
            }
            Message::ToolResult { tool_call_id, content } => {
                // Anthropic: tool results go in a "user" message with tool_result blocks
                let text = content
                    .iter()
                    .filter_map(|p| match p {
                        ContentPart::Text(t) => Some(t.as_str()),
                        _ => None,
                    })
                    .collect::<Vec<_>>()
                    .join("\n");
                Some(serde_json::json!({
                    "role": "user",
                    "content": [{
                        "type": "tool_result",
                        "tool_use_id": tool_call_id,
                        "content": text
                    }]
                }))
            }
        })
        .collect()
}

fn serialize_content_parts(parts: &[ContentPart]) -> serde_json::Value {
    if parts.len() == 1 {
        if let ContentPart::Text(text) = &parts[0] {
            return serde_json::json!(text);
        }
    }
    let blocks: Vec<serde_json::Value> = parts
        .iter()
        .map(|p| match p {
            ContentPart::Text(text) => serde_json::json!({ "type": "text", "text": text }),
            ContentPart::Image { data_uri } => {
                // Anthropic uses base64 source with media_type
                // For data URIs, extract the base64 data
                if let Some(rest) = data_uri.strip_prefix("data:image/jpeg;base64,") {
                    serde_json::json!({
                        "type": "image",
                        "source": {
                            "type": "base64",
                            "media_type": "image/jpeg",
                            "data": rest
                        }
                    })
                } else {
                    serde_json::json!({ "type": "text", "text": "[image]" })
                }
            }
        })
        .collect();
    serde_json::json!(blocks)
}

fn serialize_tools(tools: &[ToolDef]) -> Vec<serde_json::Value> {
    let len = tools.len();
    tools
        .iter()
        .enumerate()
        .map(|(i, t)| {
            let mut tool = serde_json::json!({
                "name": t.name,
                "description": t.description,
                "input_schema": t.parameters,
            });
            // Cache breakpoint on last tool — caches system + all tools prefix
            if i == len - 1 {
                tool["cache_control"] = serde_json::json!({"type": "ephemeral"});
            }
            tool
        })
        .collect()
}

File: typoon/llm/anthropic.py
"""Anthropic provider — native SDK adapter.

Serializes IR → Anthropic messages format, calls via `anthropic` SDK,
parses response back to CallResponse IR. Supports prompt caching + streaming.
"""

from __future__ import annotations

import json

import anthropic

from ._retry import parse_retry_after_header, with_retry
from .ir import (
    CallResponse,
    ContentPart,
    Message,
    Role,
    ToolCallMsg,
    ToolDef,
)

_MAX_TOKENS = 64_000


def _parse_retry_after(exc: BaseException) -> float | None:
    resp = getattr(exc, "response", None)
    headers = getattr(resp, "headers", None)
    return parse_retry_after_header(headers)


def _is_retryable(exc: BaseException) -> bool:
    if isinstance(exc, (
        anthropic.RateLimitError,
        anthropic.APITimeoutError,
        anthropic.APIConnectionError,
        anthropic.InternalServerError,
    )):
        return True
    if isinstance(exc, anthropic.APIStatusError):
        status = getattr(exc, "status_code", None)
        if status == 429 or (status is not None and 500 <= status < 600):
            return True
    return False


class AnthropicProvider:
    """Anthropic native provider with prompt caching + streaming."""

    def __init__(
        self,
        *,
        api_key: str | None = None,
        base_url: str | None = None,
        model: str = "claude-sonnet-4-20250514",
        max_tokens: int = _MAX_TOKENS,
    ) -> None:
        kwargs: dict = {}
        if api_key:
            kwargs["api_key"] = api_key
        if base_url:
            kwargs["base_url"] = base_url
        self._client = anthropic.AsyncAnthropic(**kwargs)
        self._model = model
        self._max_tokens = max_tokens

    def _build_kwargs(self, messages: list[Message], tools: list[ToolDef]) -> dict:
        system = _extract_system(messages)
        api_messages = _serialize_messages(messages)
        api_tools = _serialize_tools(tools)

        kwargs: dict = {
            "model": self._model,
            "max_tokens": self._max_tokens,
            "messages": api_messages,
        }
        if system is not None:
            kwargs["system"] = system
        if api_tools:
            kwargs["tools"] = api_tools
        return kwargs

    async def call(self, messages: list[Message], tools: list[ToolDef]) -> CallResponse:
        import json
        kwargs = self._build_kwargs(messages, tools)
        response = await with_retry(
            lambda: self._client.messages.create(**kwargs),
            is_retryable=_is_retryable,
            parse_retry_after=_parse_retry_after,
            provider="anthropic",
        )

        tool_calls: list[ToolCallMsg] = []
        text_parts: list[str] = []

        for block in response.content:
            if block.type == "text":
                text_parts.append(block.text)
            elif block.type == "tool_use":
                tool_calls.append(ToolCallMsg(
                    id=block.id,
                    name=block.name,
                    arguments=json.dumps(block.input),
                ))

        text = "".join(text_parts) if text_parts else None
        return CallResponse(tool_calls=tool_calls, text=text)


# ── IR → Anthropic serialization ─────────────────────────────────────


def _extract_system(messages: list[Message]) -> list[dict] | None:
    """Extract system message with cache_control for prompt caching."""
    for msg in messages:
        if msg.role == Role.SYSTEM and msg.text:
            return [{"type": "text", "text": msg.text, "cache_control": {"type": "ephemeral"}}]
    return None


def _serialize_messages(messages: list[Message]) -> list[dict]:
    result = []
    for msg in messages:
        match msg.role:
            case Role.SYSTEM:
                continue  # handled separately
            case Role.USER:
                result.append({"role": "user", "content": _serialize_parts(msg.parts)})
            case Role.ASSISTANT:
                blocks: list[dict] = []
                if msg.text:
                    blocks.append({"type": "text", "text": msg.text})
                for tc in msg.tool_calls:
                    inp = json.loads(tc.arguments) if tc.arguments else {}
                    blocks.append({
                        "type": "tool_use",
                        "id": tc.id,
                        "name": tc.name,
                        "input": inp,
                    })
                result.append({"role": "assistant", "content": blocks})
            case Role.TOOL_RESULT:
                text = "\n".join(p.text for p in msg.parts if p.text)
                result.append({
                    "role": "user",
                    "content": [{
                        "type": "tool_result",
                        "tool_use_id": msg.tool_call_id,
                        "content": text,
                    }],
                })
    return result


def _serialize_parts(parts: list[ContentPart]) -> str | list[dict]:
    if len(parts) == 1 and parts[0].text is not None and parts[0].image_data_uri is None:
        return parts[0].text
    blocks: list[dict] = []
    for p in parts:
        if p.text is not None:
            blocks.append({"type": "text", "text": p.text})
        if p.image_data_uri is not None:
            data_uri = p.image_data_uri
            if data_uri.startswith("data:image/jpeg;base64,"):
                b64 = data_uri[len("data:image/jpeg;base64,"):]
                blocks.append({
                    "type": "image",
                    "source": {"type": "base64", "media_type": "image/jpeg", "data": b64},
                })
            elif data_uri.startswith("data:image/png;base64,"):
                b64 = data_uri[len("data:image/png;base64,"):]
                blocks.append({
                    "type": "image",
                    "source": {"type": "base64", "media_type": "image/png", "data": b64},
                })
            else:
                blocks.append({"type": "text", "text": "[unsupported image format]"})
    return blocks


def _serialize_tools(tools: list[ToolDef]) -> list[dict]:
    n = len(tools)
    result = []
    for i, t in enumerate(tools):
        tool: dict = {
            "name": t.name,
            "description": t.description,
            "input_schema": t.parameters,
        }
        # Cache breakpoint on last tool — caches system + all tools prefix
        if i == n - 1:
            tool["cache_control"] = {"type": "ephemeral"}
        result.append(tool)
    return result
